src/pudl_scrapers/bin/ferc_xbrl.py

```diff
@@ -0,0 +1,372 @@
+    taxonomy.
+
+    Args:
+        form: Ferc form.
+        year: Year of taxonomy version.
+        archive: zipfile context manager.
+    """
+    # Get date used in entry point URL (first day of year)
+    date = datetime.date(year, 1, 1)
+
+    # Get integer form number
+    form_number = form.as_int()
+
+    logger.info(f"Archiving ferc{form_number}-{year} taxonomy")
+
+    # Construct entry point URL
+    taxonomy_entry_point = f"https://ecollection.ferc.gov/taxonomy/form{form_number}/{date.isoformat()}/form/form{form_number}/form-{form_number}_{date.isoformat()}.xsd"
+
+    # Use Arelle to parse taxonomy
+    cntlr = Cntlr.Cntlr()
+    cntlr.startLogging(logFileName="logToPrint")
+    model_manager = ModelManager.initialize(cntlr)
+    taxonomy = ModelXbrl.load(model_manager, taxonomy_entry_point)
+
+    # Loop through all files and save to appropriate location in archive
+    for url in taxonomy.urlDocs.keys():
+        url = urlparse(url)
+
+        # There are some generic XML/XBRL files in the taxonomy that should be skipped
+        if not url.netloc.endswith("ferc.gov"):
+            continue
+
+        # Download file
+        file = requests.get(url.geturl())
+
+        path = Path(url.path).relative_to("/")
+
+        with archive.open(str(path), "w") as f:
+            f.write(file.text.encode("utf-8"))
+
+
+def archive_year(year: Year, filings: set[FeedEntry], form: FercForm, output_dir: Path):
+    """Archive a single year of data for a desired form.
+
+    Args:
+        year: Year to archive.
+        filings: Set of filings indexed from RSS feed.
+        form: Ferc form.
+        output_dir: Directory to save archived filings in.
+    """
+    # Get form number as integer
+    form_number = form.as_int()
+
+    metadata = {}
+    archive_path = output_dir / f"ferc{form_number}-xbrl-{year}.zip"
+
+    with zipfile.ZipFile(archive_path, "w") as archive:
+        # Archive taxonomy
+        # The first version of the taxonomy was released in 2020
+        if year > 2019:
+            archive_taxonomy(form, year, archive)
+
+        for filing in filings:
+            # Download filing
+            xbrl_file = requests.get(filing.download_url)
+
+            # Add filing metadata
+            filing_name = f"{filing.title}{filing.ferc_period}"
+            if filing_name in metadata:
+                metadata[filing_name].update({filing.entry_id: filing.json()})
+            else:
+                metadata[filing_name] = {filing.entry_id: filing.json()}
+
+            # Write to zipfile
+            with archive.open(f"{filing.entry_id}.xbrl", "w") as f:
+                logger.info(f"Writing {filing.title} to form {form_number} archive.")
+                f.write(xbrl_file.text.encode("utf-8"))
+
+        # Save snapshot of RSS feed
+        with archive.open("rssfeed", "w") as f:
+            logger.info("Writing rss feed metadata to archive.")
+            f.write(json.dumps(metadata).encode("utf-8"))
```

Comment:
This step took a couple of minutes to complete even though the file is only a KB. Maybe something to do with the multiprocessing?

Reply:
I just reworked the multiprocessing a bit, and it seems to be working much better locally now

Reply:
How long does scraping Form 1 take for you?

Reply:
It still takes several minutes in total. Unfortunately since there's no bulk download option, and we have to download each filing individually, we ultimately have to download thousands of files in total.